src/daemon/daemon.cpp

```diff
@@ -296,6 +350,23 @@ auto try_mem_size(const std::string& val) -> mp::optional<mp::MemorySize>
     }
 }
 
+std::vector<mp::NetworkInterface> validate_extra_interfaces(const mp::LaunchRequest* request,
+                                                            mp::LaunchError& option_errors)
+{
+    // The associated bool indicates whether the interface mode is 'auto', i.e., needs to be configured by cloud-init.
+    std::vector<mp::NetworkInterface> interfaces;
+
+    for (const auto& net : request->network_options())
+    {
+        if (const auto& mac = net.mac_address(); mac.empty() || mpu::valid_mac_address(mac))
+            interfaces.emplace_back(
+                mp::NetworkInterface{net.id(), mac, net.mode() != multipass::LaunchRequest_NetworkOptions_Mode_MANUAL});
```

Comment:
The idea with `emplace*`, if for the object to be constructed in place. So, the caller provides the args and the container calls the ctor, as opposed to the caller contructing a temporary that the container copies/moves (e.g. with `push_back`). In here, you are creating a `NetworkInterface` object explicitly, so that is the argument you are calling `emplace_back` with, which means it will then call either the copy or the move constructor (depending on whether the vector as to grow IIRC).

```suggestion
            interfaces.emplace_back(net.id(), mac, net.mode() != multipass::LaunchRequest_NetworkOptions_Mode_MANUAL);
```

Reply:
Oh actually this won't compile until [C++20](https://stackoverflow.com/a/61205386). I've fallen for this before... :unamused: So the right thing would be to either provide the constructor explicitly or call `push_back` instead (where the compiler may optimize the copy away).

Reply:
So just this instead:

```suggestion
            interfaces.push_back(
                mp::NetworkInterface{net.id(), mac, net.mode() != multipass::LaunchRequest_NetworkOptions_Mode_MANUAL});
```

Reply:
Ok, thanks for the explanation. Indeed, your first suggestion `emplace_back(...)` compiles under Linux (g++ and clang) and Windows. That is used in other places of the implementation (see for instance https://github.com/canonical/multipass-private/blob/control-network-interfaces-no-driver-specific-matching/src/iso/cloud_init_iso.cpp#L442-#L443 ), so I think it is safe to do as we did in other places. What do you think?

Reply:
Oops, I was completely wrong: it doesn't compile. So following your second suggestion :(

Reply:
Hehe, you went through the same process that I did :smile: Yeah, it is normally the right thing to do, but for that the type needs to have a constructor matching the args. 

`NetworkInterface` is an aggregate, so it is considered a [POD](https://en.cppreference.com/w/cpp/named_req/PODType) and construction is different (you can initialize with `{a,b,c}` but it is aggregate initialization, similar to initializing an array). The types in the example you linked are not PODs and they have the necessary ctors. 

IOW, we can't always say `emplace_back(a,b,c)` just because we can say `T{a,b,c}`. It appears to be an oversight that they're fixing with C++20.